gr-digital/lib/ofdm_cyclic_prefixer_impl.cc

```
@@ -31,135 +31,147 @@ namespace gr {
+      : gr::tagged_stream_block
+        (
+          "ofdm_cyclic_prefixer",
+          gr::io_signature::make(1, 1, fft_len*sizeof(gr_complex)),
+          gr::io_signature::make(1, 1, sizeof(gr_complex)), len_tag_key
+        ),
+        d_fft_len(fft_len),
+        d_state(0),
+        d_cp_max(0),
+        d_rolloff_len(rolloff_len),
+        d_cp_lengths(cp_lengths),
+        d_up_flank((rolloff_len ? rolloff_len-1 : 0), 0),
+        d_down_flank((rolloff_len ? rolloff_len-1 : 0), 0),
+        d_delay_line(0, 0),
+        d_len_tag_key(len_tag_key)
     {
-      set_relative_rate(d_output_size);
-
-      // Flank of length 1 would just be rectangular
-      if (d_rolloff_len == 1) {
-	d_rolloff_len = 0;
+      for(const int& cp_len : d_cp_lengths) {
+        d_cp_max = std::max(d_cp_max, cp_len);
       }
-      if (d_rolloff_len) {
-	d_delay_line.resize(d_rolloff_len-1, 0);
-	if (rolloff_len > d_cp_size) {
-	  throw std::invalid_argument("cyclic prefixer: rolloff len must smaller than the cyclic prefix.");
-	}
-	// The actual flanks are one sample shorter than d_rolloff_len, because the
-	// first sample of the up- and down flank is always zero and one, respectively
-	for (int i = 1; i < d_rolloff_len; i++) {
-	  d_up_flank[i-1]   = 0.5 * (1 + cos(M_PI * i/rolloff_len - M_PI));
-	  d_down_flank[i-1] = 0.5 * (1 + cos(M_PI * (rolloff_len-i)/rolloff_len - M_PI));
-	}
+      // Sanity
+      for(unsigned i=0; i<d_cp_lengths.size(); i++) {
+        if (d_cp_lengths[i] != 0) {
+          break;
+        }
+        if (i == d_cp_lengths.size()-1) {
+          throw std::invalid_argument(this->alias() + std::string(": It does not make sense to only have one or a couple of CPs with length 0."));
+        }
       }
-
-      if (len_tag_key.empty()) {
-	set_output_multiple(d_output_size);
-      } else {
-	set_tag_propagation_policy(TPP_DONT);
+      if (d_cp_lengths.empty()) {
```

Please move this check above the previous. It doesn't make a huge difference, but it's a good practice to do these checks in the order in which they fail fastest.